src/lir.rs

```diff
@@ -0,0 +1,465 @@
+use crate::ast::{Expression, LogicalExpression, Predicate};
+
+#[derive(Debug)]
+pub struct LirProgram {
+    pub(crate) instructions: Vec<LirInstruction>,
+}
+
+impl LirProgram {
+    pub fn new() -> Self {
+        Self {
+            instructions: Vec::new(),
+        }
+    }
+}
+
+impl Default for LirProgram {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[derive(Debug)]
+pub enum LirInstruction {
+    LogicalOperator(LirLogicalOperators),
+    Predicate(Predicate),
+}
+
+impl LirInstruction {
+    pub fn as_predicate(&self) -> &Predicate {
+        match &self {
+            LirInstruction::LogicalOperator(_ops) => {
+                panic!("Call as_predicate on LogicalOperator Operand, LirProgram is wrong.")
+            }
+            LirInstruction::Predicate(p) => p,
+        }
+    }
+}
+
+#[inline]
+pub(crate) fn is_operator(instruction: &LirInstruction) -> bool {
+    match instruction {
+        LirInstruction::LogicalOperator(_op) => true,
+        LirInstruction::Predicate(_p) => false,
+    }
+}
+
+#[derive(Debug)]
+pub enum LirLogicalOperators {
```

Comment:
Not using plural form is more appropriate.
```suggestion
pub enum LirLogicalOperator {
```

Reply:
fixed.